cranelift/codegen/src/isa/s390x/inst.isle

```diff
@@ -1284,6 +1296,8 @@
 ;; implementation detail by helpers that preserve the SSA facade themselves.
 (decl emit (MInst) Unit)
 (extern constructor emit emit)
+(decl emit_safepoint (MInst) Unit)
+(extern constructor emit_safepoint emit_safepoint)
```

Comment:
And this should be in the prelude as well.

Reply:
I see that `emit` actually isn't in the prelude, but we do *use* it from the prelude. So we should probably move `emit` to the prelude as well, since in practice it has to be the same declaration across all backends anyways. Happy to have that as part of this updated PR or we can do it in another PR.

Reply:
Yes, that's why I didn't put it in the prelude ...

I agree it makes sense to move emit / emit_safepoint there, but then we should also move the implementations to machinst/isle.re, right?

I'd be happy to do that, but I think it would be better to do it in another PR, as it will touch all back-ends.

Reply:
Ah, I see.   But even if we don't move the implementaton, once the `decl` is in the prelude as `extern constructor`, *every* back-end *must* implement it or it won't compile any more.   So we'll still need to touch all back-ends.  (In fact, with the x86 `mov_mitosis` implementation, I'm not sure exactly how the safepoint variant would work.  I guess only the last instruction in the list should be the safepoint?)

Reply:
> once the `decl` is in the prelude as `extern constructor`, _every_ back-end _must_ implement it or it won't compile any more.

Correct. (Also, it is already the case that every backend must implement it or the ISLE won't compile because we are calling `emit` and assuming it is declared from inside the prelude already, it just so happens that we have N identical declarations for each backend instead of a single shared declaration in the prelude.)

> (In fact, with the x86 `mov_mitosis` implementation, I'm not sure exactly how the safepoint variant would work. I guess only the last instruction in the list should be the safepoint?)

Yeah, I will have to think a bit about how to do this best. I think making only the last instruction a safepoint would work, but we might want to have some debug asserts in there too. Planning on tackling this after I'm done with my newtype wrappers PR by which time this PR will have merged.

Reply:
> I think making only the last instruction a safepoint would work,

Yeah, I think that's the correct solution -- the expanded list should always have the form of zero or more moves, followed by the original instruction; and the safepoint bit applies to the latter.

Reply:
There are some instructions that always generate their results into a particular register, and so when our SSA-ified version of that instruction wants the result in a temp, we have to generate a move that follows the "real" instruction, so we can have trailing moves as well after move mitosis. This shouldn't be the case for safepoints I don't think (except maybe ABI/calling convention stuff where results are always in `rax`?) and we should be able to just debug assert this.